gpu_memory_management/test/test.hpp

```diff
@@ -0,0 +1,142 @@
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <cmath>
+#include <cstdio>
+#include <time.h>
+#include <iostream>
+#include <complex>
+#include <string>
+#include <random>
+
+#define CUDA_CHECK(call) {cudaError_t error = call; if(error!=cudaSuccess){printf("<%s>:%i ",__FILE__,__LINE__); printf("[CUDA] Error: %s\n", cudaGetErrorString(error));}}
+using namespace std;
+using namespace std::literals::complex_literals;
+namespace py = pybind11;
+
+//test if a pointer is a device/gpu pointer or not
+int is_device_pointer(const void *ptr)
```

Comment:
For some of your tests, you can use the return value `bool`, because the test result can be only correct or not. This allows also to write more clean code, e.g.:

```c++
bool is_device_pointer(const void *ptr){
  cudaPointerAttributes attributes;

  CUDA_CHECK(cudaPointerGetAttributes(&amp;attributes, ptr));

  //according to documentation, if cudaMemoryTypeDevice = 2, then it is a device memory
  // will return true, if device memory
  return attributes.type == 2;
}
```